### ISSUE TITLE
Enable running custom plugins without an admin kubeconfig

### ISSUE THREAD

**Issue by tanmay-g** [vmware-tanzu/sonobuoy#1470]
**Describe the solution you'd like**
We (tagging @janipee) have a set of tests that we have added to custom plugins for sonobuoy. The tests themselves can be run with e.g. the default `edit` ClusterRole. These tests would need to be run in clusters where we do not have cluster-admin access (RBAC is enabled), and certain namespaces would be created for our use beforehand (i.e. we cannot create and delete namespaces).

Sonobuoy currently lists an admin `kubeconfig` in its prerequisites, but assuming that the only tests/plugins to be run are the custom ones mentioned previously (i.e. e2e and systemd-logs can be ignored), would it be feasible to introduce a non-admin mode to enable runs on locked-down clusters?

E.g. currently sonobuoy attempts to create its own namespace along with a ClusterRole containing
```
rules:
- apiGroups:
  - '*'
  resources:
  - '*'
  verbs:
  - '*'
- nonResourceURLs:
  - '/metrics'
  - '/logs'
  - '/logs/*'
  verbs:
  - 'get'
```
Could this be made dependent on something like an `adminMode` parameter?

**Anything else you would like to add:**
If this proposal sounds reasonable, we could work on it and submit a PR.

**Environment:**

- Sonobuoy version: v0.53.2
- Kubernetes version: (use `kubectl version`): Varies - v1.20 onwards
- Kubernetes installer & version: Varies
- Cloud provider or hardware configuration: Varies
- OS (e.g. from `/etc/os-release`): Varies

**Comment by NikhilSharmaWe** [comment]
> 💯 You read my mind @tanmay-g
> 
> Just yesterday I was thinking about that section of the config and think this is a great idea. Currently, you can use `sonobuoy gen > out.yaml; <edit permissions>; sonobuoy run -f out.yaml` to get the same effect but I agree that isn't a great flow.
> 
> I have no problem making that configurable via a flag. Just like the recent PR I did on #1449 you could make a flag for `--aggregator-permissions=[admin| ...]`
> 
> We can talk in the PR about what the name/level should be for the non-admin mode though.
> 
> Maybe one for readOnly in the cluster, along with stronger permissions in its own namespace?

I would like to contribute, can I make a PR making a new flag. I am energized to learn how we will approach to make appropriate changes to improve the functionality.

**Comment by NikhilSharmaWe** [comment]
I have not added any functionality because I am just beginning to learn about this project. I want to see how we will go further in this issue to add this flag.

**Comment by johnSchnake** [comment]
You can take a look at the delete logic; it sounds to me like thats a separate issue. And `--delete` already has an `--all` flag to clean up other e2e namespaces so maybe we just need a simple `--keep-namespace` flag or something like that.

I don't think it lines up with the target of this ticket which is about the limiting of required permissions to launch test runs.

**Comment by johnSchnake** [comment]
@tanmay-g So I'm back to work and wanting to solve this issue for y'all. I had a question though:

In your use case, would it be acceptable to have permissions that require the ability to query for nodes? The getting of nodes requires a cluster role, but is a long way from the current cluster admin Sonobuoy currently assumes.

The reason I ask is that Sonobuoy will query the nodes to deploy and monitor daemonset plugins and even the e2e plugin itself (in the upstream logic) will need to check the nodes to ensure the cluster is ready for testing.

**Comment by tanmay-g** [comment]
@johnSchnake Sorry for the delayed reply.

Unfortunately the platform we're targeting won't allow the permissions you've described - we will be strictly restricted to our namespace.
In fact, in some cases we might not even have permissions to deploy daemonsets - but that shouldn't affect sonobuoy other than preventing the systemd-logs and e2e plugins from being run, right?

**Comment by johnSchnake** [comment]
The systemd-logs is a daemonset plugin, but the e2e plugin is not.

However, the inability to query nodes _may_ impede the e2e plugin from running due to the startup logic. I'll check though because there may be some workaround there. Once you get past that though, I'm not sure if any of the e2e tests will run at all though since I think those also require the ability to create namespaces so that they tests are sandboxed from each other.

Thanks for the response though.

**Comment by johnSchnake** [comment]
I'll have a think about what we can do about the logic for the node query inside of Sonobuoy.

I think that should be easy enough to work around; I think a reasonable way to approach it in such a limited way is to just limit Sonobuoy's functionality in that case. If we can't query nodes we either say daemonset plugins are unable to be run and/or we can't monitor them in the same way (not sure if that is achievable though since we wouldn't even know how may results we would be waiting on or when we know we'd be done).

**Comment by tanmay-g** [comment]
> the inability to query nodes may impede the e2e plugin

At least for our particular use-case, it is completely fine to declare that the 'namespaced' mode does not support e2e, systemd-logs, or anything else requiring cluster-scope permissions.
I.e. when in the 'namespaced' mode, the only thing we need to run are our custom plugins (which responsible to stay namespaced).

The cluster admin can choose to run in 'admin/non-namespaced' mode (i.e. with full cluster-admin permissions) to run the other plugins (e.g. e2e, which needs its own namespaces).
